### ISSUE TITLE
vo_drm: Playback "tearing" when video doesn't match monitor's aspect ratio

### ISSUE THREAD

**Issue by robin007bond** [mpv-player/mpv#1830]
The vo_drm driver works perfectly here when playing videos that have the same aspect ratio as my monitor (16:9). However, when I play videos with an aspect ratio of 1.333, which doesn't fill my screen, a lot of tearing occurs in the lower 25% of the screen.

I tried several videos and they work fine with the open-gl drivers. Playing back other videos in 16:9 works perfectly (I tried several files), but videos in other aspect ratios don't work perfectly.

By the way, the video driver itself is great! I was amazed at how good the playback was, even with 10-bit encoded HD content.

**Comment by rr-** [comment]
I'll fix more important bugs first (#1827, #1828) and then I'll get to tearing.

**Comment by robin007bond** [comment]
I understand. Nevertheless, great work you're doing on vo-drm. :+1:

**Comment by rr-** [comment]
Alright. @robin007bond are you sure you experience no problems whatsoever with all 16:9 videos?
It seems unlikely to me that aspect ratio has something to do with this.

Could you test if scaling your video with `-vf` to your screen's native resolution helps?

**Comment by robin007bond** [comment]
Hmm.. I've only tested a few 16:9 videos. I will check if I can test more and look if the tearing stops when scaling the 4:3 videos to my native resolution.

**Comment by robin007bond** [comment]
I've tested another 16:9 video and noticed mild tearing when the video pans upwards or downwards. 

Some info on the tested files and whether tearing is present:

Tearing present when panning downwards or upwards:
```
Video
ID                                       : 1
Format                                   : AVC
Format/Info                              : Advanced Video Codec
Format profile                           : High@L5
Format settings, CABAC                   : Yes
Format settings, ReFrames                : 11 frames
Codec ID                                 : V_MPEG4/ISO/AVC
Duration                                 : 24mn 27s
Width                                    : 1 280 pixels
Height                                   : 720 pixels
Display aspect ratio                     : 16:9
Frame rate mode                          : Constant
Frame rate                               : 23.976 fps
Color space                              : YUV
Chroma subsampling                       : 4:2:0
Bit depth                                : 8 bits
Scan type                                : Progressive
Title                                    : [H264]
Writing library                          : x264 core 112 r1867 22bfd31
Encoding settings                        : cabac=1 / ref=12 / deblock=1:1:1 / analyse=0x3:0x113 / me=umh / subme=10 / psy=1 / psy_rd=1.00:0.00 / mixed_ref=1 / me_range=24 / chroma_me=1 / trellis=2 / 8x8dct=1 / cqm=0 / deadzone=21,11 / fast_pskip=1 / chroma_qp_offset=-4 / threads=12 / sliced_threads=0 / nr=0 / decimate=1 / interlaced=0 / constrained_intra=0 / bframes=12 / b_pyramid=1 / b_adapt=2 / b_bias=0 / direct=2 / weightb=1 / open_gop=0 / weightp=2 / keyint=250 / keyint_min=23 / scenecut=40 / intra_refresh=0 / rc=crf / mbtree=0 / crf=17.0 / qcomp=0.60 / qpmin=10 / qpmax=51 / qpstep=4 / ip_ratio=1.40 / pb_ratio=1.30 / aq=2:0.50
Language                                 : Japanese
Default                                  : Yes
Forced                                   : No
```
Tearing very noticeable, and also video shaking a bit (it's hard to explain this behavior). This is not the case with vo_opengl. I've stretched the aspect ratio with `--aspect-ratio=16:9`, and it didn't reduce tearing. Using `--vf=scale:1920:1080` didn't work either. 
```
Video
ID                                       : 0
Format                                   : MPEG-4 Visual
Format profile                           : Simple@L3
Format settings, BVOP                    : 1
Format settings, QPel                    : No
Format settings, GMC                     : No warppoints
Format settings, Matrix                  : Default (H.263)
Codec ID                                 : DIVX
Codec ID/Info                            : Project Mayo
Codec ID/Hint                            : DivX 4
Duration                                 : 24mn 9s
Bit rate                                 : 870 Kbps
Width                                    : 640 pixels
Height                                   : 480 pixels
Display aspect ratio                     : 4:3
Frame rate                               : 23.976 fps
Color space                              : YUV
Chroma subsampling                       : 4:2:0
Bit depth                                : 8 bits
Scan type                                : Progressive
Compression mode                         : Lossy
Bits/(Pixel*Frame)                       : 0.118
Stream size                              : 150 MiB (86%)
Writing library                          : XviD 0.0.09 (UTC 2003-03-25)

```
Working perfectly (even no tearing in scenes with panning)
```
Video
ID                                       : 1
Format                                   : AVC
Format/Info                              : Advanced Video Codec
Format profile                           : High 10@L5
Format settings, CABAC                   : Yes
Format settings, ReFrames                : 16 frames
Codec ID                                 : V_MPEG4/ISO/AVC
Duration                                 : 23mn 40s
Width                                    : 1 280 pixels
Height                                   : 720 pixels
Display aspect ratio                     : 16:9
Frame rate mode                          : Constant
Frame rate                               : 23.976 fps
Color space                              : YUV
Chroma subsampling                       : 4:2:0
Bit depth                                : 10 bits
Scan type                                : Progressive
Writing library                          : x264 core 142 r2431+42 c69a006 tMod [10-bit@all X86_64]
Encoding settings                        : cabac=1 / ref=16 / deblock=1:0:0 / analyse=0x3:0x133 / me=umh / subme=10 / psy=1 / fade_compensate=0.70 / psy_rd=0.60:0.15 / mixed_ref=1 / me_range=24 / chroma_me=1 / trellis=2 / 8x8dct=1 / cqm=0 / deadzone=21,11 / fast_pskip=1 / chroma_qp_offset=-3 / threads=12 / lookahead_threads=3 / sliced_threads=0 / nr=0 / decimate=1 / interlaced=0 / bluray_compat=0 / constrained_intra=0 / fgo=0 / bframes=10 / b_pyramid=2 / b_adapt=2 / b_bias=0 / direct=3 / weightb=1 / open_gop=0 / weightp=2 / keyint=250 / keyint_min=23 / scenecut=40 / intra_refresh=0 / rc_lookahead=60 / rc=crf / mbtree=1 / crf=18.0000 / qcomp=0.70 / qpmin=0:0:0 / qpmax=81:81:81 / qpstep=4 / ip_ratio=1.40 / aq=3:0.80 / aq-sensitivity=10.00 / aq-factor=1.00:1.00:1.00 / aq2=0 / aq3=0
Language                                 : English
Default                                  : Yes
Forced                                   : No
DURATION                                 : 00:23:40.044000000
NUMBER_OF_FRAMES                         : 34047
NUMBER_OF_BYTES                          : 233245575
_STATISTICS_WRITING_APP                  : mkvmerge v7.8.0 ('River Man') 64bit built on Mar 27 2015 16:31:37
_STATISTICS_WRITING_DATE_UTC             : 2015-04-11 18:37:48
_STATISTICS_TAGS                         : BPS DURATION NUMBER_OF_FRAMES NUMBER_OF_BYTES

```
Feel fry to suggest other files for testing.

**Comment by rr-** [comment]
Thank you for performing the tests. The only distinguishing thing I can see is that "bad" videos are 8bit, while the "good" one is 10bit. I don't think any more testing is going to reveal anything interesting, so let's not waste time on that.

@robin007bond, I'd like however to ask you to test the same videos once I implement vsync, which should (note: *should*) happen some time this week.

PS

>Duration                                 : 24mn 27s
>Duration                                 : 24mn 9s
>Duration                                 : 23mn 40s

I see what you did there :+1:

**Comment by robin007bond** [comment]
Hmm, yes. That's also the only distinguishing thing that I see. The ones with an avi container and being encoded in Xvid, perform a lot worse than the h264 files.

Out of curiosity, I tested one thing. The bad "Xvid in Avi" files perform bad too when the video and audio streams are merged in a mkv.  Also, the problem still persists when encoding the "bad" files to h264. 

Anyway, I'll test them again when vsync is implemented and will report the results back.

**Comment by rr-** [comment]
Should be fixed in a7cf35c1ca08cf59ae9f1d9e56c7fb486af6e9ff.

**Comment by robin007bond** [comment]
Yes, it is fixed.
